### DISCUSSION TITLE
Transforming pose of Node from one anchor to another anchor

### DISCUSSION THREAD

**Discussion by sarimmehdi** [sceneview/sceneform-android#277] [Q&A]
I want to compute a new pose for the node to position it on a new anchor. My initial idea was to use the pose of the first anchor and the second anchor to bring the node from the reference frame of the first anchor to second anchor:

```
// get pose of node in reference frame of first anchor
Vector3 nodeLocalPosition = node.getLocalPosition();
Quaternion nodeLocalRotation = node.getLocalRotation();
float[] t = new float[]{nodeLocalPosition.x, nodeLocalPosition.y, nodeLocalPosition.z};
float[] r = new float[]{nodeLocalRotation.y, nodeLocalRotation.y, nodeLocalRotation.z, nodeLocalRotation.w};
Pose nodeToFirstAnchor = new Pose(t, r);

// get pose of first anchor in world reference frame
Pose firstAnchorToWorld = firstAnchor.getPose();

// get pose of second anchor in world reference frame
Pose secondAnchorToWorld = secondAnchor.getPose();
Pose worldToSecondAnchor = secondAnchorToWorld.inverse();

// transform node from reference frame to first anchor to reference frame of second anchor
Pose nodeToSecondAnchor = worldToSecondAnchor.compose(firstAnchorToWorld.compose(nodeToFirstAnchor));
// and then I set the local position and rotation of the node in the second anchor and make the second anchor the parent of this node
```
Then I noticed that sceneform provides world position and rotation of the node as well. So, I decided to use those:

```
// get pose of node in world reference frame instead
Vector3 nodeWorldPosition = node.getWorldPosition();
Quaternion nodeWorldRotation = node.getWorldRotation();
float[] t = new float[]{nodeWorldPosition.x, nodeWorldPosition.y, nodeWorldPosition.z};
float[] r = new float[]{nodeWorldRotation.y, nodeWorldRotation.y, nodeWorldRotation.z, nodeWorldRotation.w};
Pose nodeToWorld = new Pose(t, r);

// get pose of second anchor in world reference frame (notice I am not using the pose of the first anchor since I needed that before to bring the node into the world reference frame but now I am able to get the node's pose in the world reference frame directly thanks to Sceneform)
Pose secondAnchorToWorld = secondAnchor.getPose();
Pose worldToSecondAnchor = secondAnchorToWorld.inverse();

// transform node from reference frame to first anchor to reference frame of second anchor
Pose nodeToSecondAnchor = worldToSecondAnchor.compose(nodeToWorld);
```

Now, my question is, how does sceneform compute the pose of a simple node in the world coordinate frame? Does it use the pose of the camera and anchors to recompute the pose of the node? I tried following the logic [here](https://github.com/SceneView/sceneform-android/blob/ba2c818151b91296f048711f494e1acc239ff809/core/src/main/java/com/google/ar/sceneform/Node.java#L640) but there are a lot of keywords that don't really make sense (WORLD_POSITION_DIRTY, LOCAL_TRANSFORM_DIRTY etc.). Is the process of recomputing the pose of the node in the world coordinate system a simple mathematical process (using the latest and previous estimated camera pose provided by ARCore to update the position of all Nodes or something along that line) or is there some computer vision stuff being used as well?

**Comment by grassydragon** [comment]
> Does it use the pose of the camera and anchors to recompute the pose of the node?

Only the parent nodes participate in calculation of the world transform. If there is a parent `AnchorNode`, when the latest anchor position is used.

> I tried following the logic here but there are a lot of keywords that don't really make sense (WORLD_POSITION_DIRTY, LOCAL_TRANSFORM_DIRTY etc.).

This is a caching logic since calculating the world transform can be expensive if there are many parent nodes.

> Is the process of recomputing the pose of the node in the world coordinate system a simple mathematical process (using the latest and previous estimated camera pose provided by ARCore to update the position of all Nodes or something along that line) or is there some computer vision stuff being used as well?

Yes, that is a simple mathematical process of multiplying the transform matrices recursively as you can see in the following method. Again, the camera transform isn't required to compute the world transform of the node.

https://github.com/SceneView/sceneform-android/blob/65d6e6ecccabd56fb21c64464793f750d60ccaf7/core/src/main/java/com/google/ar/sceneform/Node.java#L1519
